ddl/lightning/engine_mgr.go

```diff
@@ -0,0 +1,111 @@
+// limitations under the License.
+
+package lightning
+
+import (
+	"fmt"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/util/generic"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+type engineManager struct {
+	generic.SyncMap[int64, *engineInfo]
+	MemRoot  MemRoot
+	DiskRoot DiskRoot
+}
+
+func (m *engineManager) init(memRoot MemRoot, diskRoot DiskRoot) {
+	m.SyncMap = generic.NewSyncMap[int64, *engineInfo](10)
+	m.MemRoot = memRoot
+	m.DiskRoot = diskRoot
+}
+
+// Register create a new engineInfo and register it to the engineManager.
+func (m *engineManager) Register(bc *BackendContext, job *model.Job, indexID int64) error {
+	// Calculate lightning concurrency degree and set memory usage
+	// and pre-allocate memory usage for worker.
+	m.MemRoot.RefreshConsumption()
+	ok := m.MemRoot.TestConsume(int64(bc.cfg.TikvImporter.LocalWriterMemCacheSize))
+	if !ok {
+		return logAllocMemFailedEngine(m.MemRoot, bc.jobID, indexID)
+	}
+
+	en, exist1 := m.Load(indexID)
+	if !exist1 {
+		engineCacheSize := int64(bc.cfg.TikvImporter.EngineMemCacheSize)
+		ok := m.MemRoot.TestConsume(StructSizeEngineInfo + engineCacheSize)
+		if !ok {
+			return logAllocMemFailedEngine(m.MemRoot, bc.jobID, indexID)
+		}
+
+		cfg := generateLocalEngineConfig(job.ID, job.SchemaName, job.TableName)
+		openedEn, err := bc.backend.OpenEngine(bc.ctx, cfg, job.TableName, int32(indexID))
+		if err != nil {
+			return errors.New(LitErrCreateEngineFail)
+		}
+		id := openedEn.GetEngineUUID()
+		en = NewEngineInfo(bc.ctx, job.ID, indexID, cfg, openedEn, id, 1, m.MemRoot, m.DiskRoot)
+		m.Store(indexID, en)
+		if err != nil {
+			return errors.New(LitErrCreateEngineFail)
+		}
+		m.MemRoot.Consume(StructSizeEngineInfo)
+		m.MemRoot.ConsumeWithTag(encodeEngineTag(job.ID, indexID), engineCacheSize)
+	} else {
+		if en.writerCount+1 > bc.cfg.TikvImporter.RangeConcurrency {
+			logutil.BgLogger().Warn(LitErrExceedConcurrency, zap.Int64("job ID", job.ID),
+				zap.Int64("index ID", indexID),
+				zap.Int("concurrency", bc.cfg.TikvImporter.RangeConcurrency))
+			return errors.New(LitErrExceedConcurrency)
+		}
+		en.writerCount++
+	}
+	m.MemRoot.ConsumeWithTag(encodeEngineTag(job.ID, indexID), int64(bc.cfg.TikvImporter.LocalWriterMemCacheSize))
+	logutil.BgLogger().Info(LitInfoOpenEngine, zap.Int64("job ID", job.ID),
+		zap.Int64("index ID", indexID),
+		zap.Int64("current memory usage", m.MemRoot.CurrentUsage()),
+		zap.Int64("memory limitation", m.MemRoot.MaxMemoryQuota()),
+		zap.Int("current writer count", en.writerCount))
+	return nil
+}
+
+// Unregister delete the engineInfo from the engineManager.
+func (m *engineManager) Unregister(jobID, indexID int64) {
+	ei, exists := m.Load(indexID)
+	if !exists {
+		return
+	}
+
+	ei.Clean()
+	m.Delete(indexID)
+	m.MemRoot.ReleaseWithTag(encodeEngineTag(jobID, indexID))
+	m.MemRoot.Release(StructSizeWriterCtx * int64(ei.writerCount))
```

Comment:
I don't see `Consume` `StructSizeWriterCtx` in `Register`, so why do we need to release it here?

Reply:
Done. https://github.com/pingcap/tidb/pull/37521/commits/b1bfdbaeb91163231543909c07647814bda2fba2